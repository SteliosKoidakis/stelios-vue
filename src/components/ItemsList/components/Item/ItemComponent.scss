.ItemComponent {
  display: flex;
  flex-direction: column;
  height: 100%;
  padding-bottom: map-get($spacings, 3);

  &__image-wrapper {
    position: relative;
    padding-bottom: 56.25%; // aspect ration 16:9 
    background-color: lightgoldenrodyellow;
  }

  &__image {
    position: absolute;
    width: 100%;
    height: 100%;

    &[lazy="loading"] {
      opacity: 0;
    }

    &[lazy="loaded"] {
      top: 0;
      left: 0;
      max-height: 100%;
      opacity: 1;
      transition: opacity 0.8s;
    }
  }

  &__description {
    padding-top: map-get($spacings, 3);
    flex-grow: 1;
  }

  &__email {
    display: block;
    padding-bottom: map-get($spacings, 3);
    word-break: break-all;
  }
}