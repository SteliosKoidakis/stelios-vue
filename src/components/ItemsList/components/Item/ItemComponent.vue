<template>
  <div class="ItemComponent">
    <div class="ItemComponent__image-wrapper">
      <img
        v-if="image"
        v-lazy="image"
        :alt="title"
        :title="title"
        class="ItemComponent__image"
        onerror="this.style.display='none'"
      >
    </div>
    <h4 v-if="title">
      {{ title }}
    </h4>
    <p
      v-if="description"
      class="ItemComponent__description"
    >
      {{ description }}
    </p>
    <p v-if="price">
      {{ price }}
    </p>
    <a
      v-if="email"
      class="ItemComponent__email"
      :href="`mailto:${email}`"
    >
      {{ email }}
    </a>
    <BButton
      :variant="isFavorite ? 'success' : 'light'"
      @click="(!isFavorite || isUnlikeEnabled) && onFavoriteItem()"
    >
      {{ isFavorite ? dislikeText : 'Like' }}
    </BButton>
  </div>
</template>
<script>
import { BButton } from 'bootstrap-vue';

export default {
  name: 'ItemComponent',
  components: {
    BButton,
  },
  props: {
    title: {
      type: String,
      default: '',
    },
    description: {
      type: String,
      default: '',
    },
    price: {
      type: String,
      default: '',
    },
    email: {
      type: String,
      default: '',
    },
    image: {
      type: String,
      default: '',
    },
    isFavorite: {
      type: Boolean,
      default: false,
    },
    isUnlikeEnabled: {
      type: Boolean,
      default: false,
    },
    onFavoriteItem: {
      type: Function,
      default: () => ({}),
    },
  },
  computed: {
    dislikeText() {
      return this.isUnlikeEnabled ? 'Dislike' : 'Liked';
    },
  },
};
</script>
<style lang="scss" scoped>
 @import './ItemComponent';
</style>
