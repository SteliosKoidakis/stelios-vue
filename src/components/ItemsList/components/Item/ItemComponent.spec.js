import { shallowMount } from '@vue/test-utils';
import { BButton } from 'bootstrap-vue';

import ItemComponent from './ItemComponent';

const propsData = {
  image: 'image',
  title: 'title',
  description: 'description',
  price: 'price',
  email: 'email',
};

describe('Given the ItemsList component', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(ItemComponent, {
      propsData,
    });
  });

  describe('when the components is mounted', () => {
    it('should not break', () => {
      expect(wrapper).toBeDefined();
    });
    it('should render BButton component', () => {
      expect(wrapper.find(BButton).exists()).toBe(true);
    });
  });
  describe('when props pass', () => {
    it('should render the correct output', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
  describe('when props isFavorite is true', () => {
    it('should render Bbutton with success variant and Unlike text', async () => {
      await wrapper.setProps({
        isFavorite: true,
      });
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
