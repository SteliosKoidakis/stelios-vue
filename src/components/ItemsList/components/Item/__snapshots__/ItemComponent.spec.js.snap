// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the ItemsList component when props isFavorite is true should render Bbutton with success variant and Unlike text 1`] = `
<div
  class="ItemComponent"
>
  <div
    class="ItemComponent__image-wrapper"
  >
    <img
      alt="title"
      class="ItemComponent__image"
      src="image"
      title="title"
    />
  </div>
   
  <h4>
    
    title
  
  </h4>
   
  <p
    class="ItemComponent__description"
  >
    
    description
  
  </p>
   
  <p>
    
    price
  
  </p>
   
  <a
    class="ItemComponent__email"
    href="mailto:email"
  >
    
    email
  
  </a>
   
  <bbutton-stub
    event="click"
    routertag="a"
    tag="button"
    target="_self"
    type="button"
    variant="success"
  >
    
    Unlike
  
  </bbutton-stub>
</div>
`;

exports[`Given the ItemsList component when props pass should render the correct output 1`] = `
<div
  class="ItemComponent"
>
  <div
    class="ItemComponent__image-wrapper"
  >
    <img
      alt="title"
      class="ItemComponent__image"
      src="image"
      title="title"
    />
  </div>
   
  <h4>
    
    title
  
  </h4>
   
  <p
    class="ItemComponent__description"
  >
    
    description
  
  </p>
   
  <p>
    
    price
  
  </p>
   
  <a
    class="ItemComponent__email"
    href="mailto:email"
  >
    
    email
  
  </a>
   
  <bbutton-stub
    event="click"
    routertag="a"
    tag="button"
    target="_self"
    type="button"
    variant="light"
  >
    
    Like
  
  </bbutton-stub>
</div>
`;
